### PR TITLE
F#src 3375 aurora documentation

### PR DESCRIPTION
**Pull request by ToivoHn**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,129 @@
...
+
+The preferred way to develop code for this project is to pull a certain docker image with a lot of tools already installed and open a container in it, then clone the aurora GitHub repository inside it. It is not recommended to clone aurora directly on your local machine while you do development and testing.
+
+Instructions how to access the docker image and container for development, see Development Docker section below
+
+### Development Docker ###
+
+The docker images used for aurora development are [here](https://cloud.docker.com/u/shadowrobot/repository/docker/shadowrobot/aurora-molecule-devel)
+
+We are currently using the bionic tag
```
Comment (AndriyPt) [inline README.md]: Currently Xenial and Bionic are working well. So there is no difference.

File: README.md [hunk 2]
```diff
@@ -1,8 +1,129 @@
...
+1. In the docker container terminal go to /ansible/playbooks folder
+
+```
+cd /home/user/aurora/ansible/playbooks
+```
+
+2. Run the following command to execute all molecule tests with debug mode:
+
+```
+molecule --debug test --all
```
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: I have fixed this is latest commit in this branch

File: README.md [hunk 3]
```diff
@@ -1,8 +1,129 @@
...
+3. Fix any errors
+
+4. If you want to run a specific test case (in ansible/playbooks/molecule folder), use:
+
+```
+molecule --debug test -s name_of_your_scenario
+```
+5. If you want to to run Molecule in stages (create, converge, etc.), see [this](https://molecule.readthedocs.io/en/stable/usage.html) page, and do, for example:
+```
+molecule --debug create -s name_of_your_scenario
```
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: Done

File: README.md (hunk 1 above)
Comment (ToivoHn) [reply README.md]: Fixed

File: README.md [hunk 4]
```diff
@@ -1,8 +1,194 @@
...
+```
+
+3. Fix any errors
+
+4. If you want to run a specific test case (scenarios are in ansible/playbooks/molecule_docker/molecule folder), use:
+
+```
+molecule --debug test -s name_of_your_scenario
+```
+5. Often it is useful to run molecule in stages (create, converge, verify, login (if necessary), and finally destory) for better debugging (so you can inspect every stage yourself). See [this](https://molecule.readthedocs.io/en/stable/usage.html) page, and do, for example:
```
Comment (giusebar) [inline README.md]: typo: destroy

File: README.md [hunk 5]
```diff
@@ -1,8 +1,194 @@
...
+The recommended way to develop code for this project is to pull a certain docker image ([Development Docker](#development-docker)) with a lot of tools already installed and open a container of this image, then clone the aurora GitHub repository inside it. It is not recommended to clone aurora directly on your local machine while you do development and testing.
+
+## Development Docker ##
+
+The docker images used for aurora development are [here](https://cloud.docker.com/u/shadowrobot/repository/docker/shadowrobot/aurora-molecule-devel)
+
+Currently both xenial and bionic tags are working well and there is no difference.
+The rest of the document assumes the user is using bionic tag.
+
+Instructions on how to use this:
```
Comment (biofotis) [inline README.md]: maybe we can have a aurora-deploy playbook later to automate this?

File: README.md [hunk 6]
```diff
@@ -1,8 +1,194 @@
...
+1. Use Ubuntu 16.04 or Ubuntu 18.04 computer
+
+2. Install Docker (using instructions from [here](https://docs.docker.com/install/linux/docker-ce/ubuntu/))
+
+3. Run the following command in terminal to create a container for aurora development:
+
+```
+docker run -it --name aurora_dev -e DISPLAY -e QT_X11_NO_MITSHM=1 -e LOCAL_USER_ID=$(id -u) -v /var/run/docker.sock:/var/run/docker.sock -v /tmp/.X11-unix:/tmp/.X11-unix:rw shadowrobot/aurora-molecule-devel:bionic
+```
+4. Once the container has launched, clone aurora to it:
```
Comment (biofotis) [inline README.md]: clone to home directory

File: README.md [hunk 7]
```diff
@@ -1,8 +1,194 @@
...
+```
+molecule --debug create -s name_of_your_scenario
+molecule --debug converge -s name_of_your_scenario
+molecule --debug verify -s name_of_your_scenario
+molecule --degub login -s name_of_your_scenario
+molecule --debug destroy -s name_of_your_scenario
+```
+## Private docker images ##
+
+At the moment, we don't want to give molecule access to private docker hub credentials for private docker images (e.g. shadow-teleop). That is why, in every playbook.yml inside the test scenarios in the molecule_docker folder, we override the image with image="shadowrobot/dexterous-hand" for any teleop-related test scenario. When we actually deploy Aurora, the user will be asked to fill in their private Docker hub credentials.
```
Comment (giusebar) [inline README.md]: Not super clear. Each test we have uses image="shadowrobot/dexterous-hand" by default and if you want to run the tests for another scenario we need to change this and then be asked for docker hub credentials?

File: README.md [hunk 8]
```diff
@@ -1,8 +1,194 @@
...
+
+## Testing with molecule_ec2 ##
+
+Once you have written your code for aurora in your branch, and tested it locally with molecule_docker, you can test it with AWS EC2 (initiated from local), by following the steps here:
+
+## Credentials ##
```
Comment (giusebar) [inline README.md]: All those automatic section should probably be subsection of the "Testing with molecule_ec2" section.

File: README.md [hunk 9]
```diff
@@ -1,8 +1,194 @@
...
+```
+## Automatic tests ##
+
+The buildspec.yml file in the root of the project defines what AWS CodeBuild should run when a PR is created or updated or when a daily build runs. It is configured to run all tests in /ansible/playbooks/molecule_ec2 folder
+
+## Test creation ##
+
+Create test scenarios for both docker in ansible/playbooks/molecule_docker/molecule folder and for AWS EC2 in ansible/playbooks/molecule_ec2/molecule folder. For additional molecule_docker tests, copy the folder structure from other tests and modify the python .py file in tests folder.For additional molecule_ec2 tests, copy the folder structure of another EC2 test and modify the molecule.yml file inside. The EC2 tests just run the same tests as the Docker tests, but they do it in AWS EC2, using virtual machines, not Docker.
+
+## Testing or deploying on real hardware ##
```
Comment (biofotis) [inline README.md]: I would move this before tests section as its logical to first create the task and then make tests for it

File: README.md [hunk 10]
```diff
@@ -12,24 +198,177 @@ command in the console. At the bottom, there will be information similar to the
...
+
+E.g. if we want to include a particular role to install Docker, we do:
 
 ```bash
-bash <(curl -Ls bit.ly/run-aurora) docker_deploy option1=value1 option2=value2 option3=value3
+- name: Include installation/docker role
+  include_role:
+    name: installation/docker
 ```
+The other way of having dependencies in Ansible is by using the meta folder and main.yml inside the meta folder. Any tasks in meta/main.yml are run before the task/main.yml. An example of meta/main.yml:
```
Comment (biofotis) [inline README.md]: if we do it like that are we going to have issues with defaults overwirting in different computers?

File: README.md [hunk 11]
```diff
@@ -12,24 +198,177 @@ command in the console. At the bottom, there will be information similar to the
...
+    - {role: products/hand-h/deploy, when: product == 'hand_h'}
+    - {role: products/hand-e/deploy, when: product == 'hand_e'}
+    - {role: docker/aws, when: use_aws|bool}
+```
+
+## teleop_deploy ##
+
+For deploying teleop software on multiple machines (server, control-machine, client, windows-machine)
+
+How to run:
```
Comment (biofotis) [inline README.md]: The only thing I didnt like is that now we are missing the quick guide that we have clearly described on top. 
If someones wants to quickly run aurora for docker deployment or teleop he will have to go through the document to find relevant sections. 
Maybe if we can have a very short section on top to show how to run it? and reference it to the proper section for more info

File: README.md [hunk 12]
```diff
@@ -1,8 +1,194 @@
...
+aws configure
+```
+2. Paste the access key and the secret access key
+
+3. Default region name must be: eu-west-2
+
+4. Press enter on the Default format
+
+
+Then continue testing with molecule_ec2:
```
Comment (giusebar) [inline README.md]: This should probably be another section

File: README.md [hunk 13]
```diff
@@ -1,8 +1,194 @@
...
+```
+molecule --debug create -s name_of_your_scenario
+molecule --debug converge -s name_of_your_scenario
+molecule --debug verify -s name_of_your_scenario
+molecule --degub login -s name_of_your_scenario
+molecule --debug destroy -s name_of_your_scenario
+```
+## Automatic tests ##
+
+The buildspec.yml file in the root of the project defines what AWS CodeBuild should run when a PR is created or updated or when a daily build runs. It is configured to run all tests in /ansible/playbooks/molecule_ec2 folder
```
Comment (giusebar) [inline README.md]: Maybe give link to an example of buildspec.yml

File: README.md [hunk 14]
```diff
@@ -1,8 +1,194 @@
...
+
+## Test creation ##
+
+Create test scenarios for both docker in ansible/playbooks/molecule_docker/molecule folder and for AWS EC2 in ansible/playbooks/molecule_ec2/molecule folder. For additional molecule_docker tests, copy the folder structure from other tests and modify the python .py file in tests folder.For additional molecule_ec2 tests, copy the folder structure of another EC2 test and modify the molecule.yml file inside. The EC2 tests just run the same tests as the Docker tests, but they do it in AWS EC2, using virtual machines, not Docker.
```
Comment (giusebar) [inline README.md]: maybe put links to some examples of the .py file and molecule.yml file

File: README.md [hunk 15]
```diff
@@ -1,8 +1,194 @@
...
+
+## Testing or deploying on real hardware ##
+
+For debugging (not using the master branch), you can add the following immediately after playbook name (for example docker_deploy or teleop_deploy):
```
Comment (giusebar) [inline README.md]: add where? in the playbook file or in a command to run the playbook?

Comment (giusebar) [reply README.md]: also link to deploy examples would be useful?

File: README.md [hunk 16]
```diff
@@ -1,8 +1,194 @@
...
+[here](https://ansible-tips-and-tricks.readthedocs.io/en/latest/ansible/commands/#limit-to-one-or-more-hosts))
+
+For assigning input and secure input to playbook variables you can use the tags: --read-input var1, var2, var3 ... and --read-secure secure_var1, secure_var2, secure_var3 ... respectively
+
+* --read-input vars (e.g. --read-input docker_username - To allow aurora script to prompt for docker username)
+* --read-secure secure_vars (e.g. --read_secure docker_password - To allow aurora script to prompt for docker password)
+
+## Ethercat interface ##
+
+Before running the docker_deploy playbook ##
```
Comment (giusebar) [inline README.md]: i guess it's in progress here

File: README.md (hunk 5 above)
Comment (ToivoHn) [reply README.md]: Yes, a good idea. Can you please create a small user story?

File: README.md (hunk 6 above)
Comment (ToivoHn) [reply README.md]: Ok, fixed

File: README.md [hunk 17]
```diff
@@ -12,24 +199,181 @@ command in the console. At the bottom, there will be information similar to the
...
+[Docker](#docker)
+
+[Installation](#installation)
+
+[Products](#products)
+
+
+## Docker ##
+
+The docker folder contains some general roles that are used in after docker install. It has the following folders:
```
Comment (giusebar) [inline README.md]: remove in

File: README.md [hunk 18]
```diff
@@ -12,24 +199,181 @@ command in the console. At the bottom, there will be information similar to the
...
+
+aws: this is used for installing our shadow-upload.sh script and AWS customer key which uploads ROS logs to AWS. It has a dependency of installation/aws-cli
```
Comment (giusebar) [inline README.md]: make pointed list

File: README.md [hunk 19]
```diff
@@ -12,24 +199,181 @@ command in the console. At the bottom, there will be information similar to the
...
+
+docker-image: this is used for pulling the docker image (and if nvidia_docker it not 0, it will append -nvidia to the docker image before it is pulled)
```
Comment (giusebar) [inline README.md]: is not 0 (why zero?)

File: README.md [hunk 20]
```diff
@@ -12,24 +199,181 @@ command in the console. At the bottom, there will be information similar to the
...
+
+Whenever you create a role, if it will likely be used by other products (e.g. it's not very specific to your product), please put it under common roles, or if it's for installing some commonly used program, put it under installation.
+
+## Templating ##
+
+For various bash/docker/etc. scripts, it's often useful to use Jinja2 templates (.j2 files). You can read more about .j2 files [here](https://docs.ansible.com/ansible/latest/user_guide/playbooks_templating.html)). They are stored either in /products/common/resources/ or if they are specific to a particular product, they should be in that products /templates/scripts folder.
+
+## Dependencies ##
+
+There are 2 main way of including dependencies in Ansible roles. The primary way we use is the "include_role" method because it is dynamic (see [here](https://docs.ansible.com/ansible/latest/modules/include_role_module.html) for documentation).
```
Comment (giusebar) [inline README.md]: The preferred way

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: Fixed

File: README.md [hunk 21]
```diff
@@ -1,8 +1,194 @@
...
+```
+git clone https://github.com/shadow-robot/aurora.git
+```
+5. Go into the aurora folder:
+```
+cd aurora
+```
+6. Open Visual Studio Code which is already installed inside the Docker container:
+```
+code
```
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: ok, fixed

File: README.md [hunk 22]
```diff
@@ -1,8 +1,194 @@
...
+
+Molecule user guide is available [here](https://molecule.readthedocs.io/en/stable/) (Aurora is currently using Molecule 2.20.1)
+
+# Development #
+
+The recommended way to develop code for this project is to pull a certain docker image ([Development Docker](#development-docker)) with a lot of tools already installed and open a container of this image, then clone the aurora GitHub repository inside it. It is not recommended to clone aurora directly on your local machine while you do development and testing.
+
+## Development Docker ##
+
+The docker images used for aurora development are [here](https://cloud.docker.com/u/shadowrobot/repository/docker/shadowrobot/aurora-molecule-devel)
```
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: fixed

File: README.md [hunk 23]
```diff
@@ -1,8 +1,194 @@
...
+```
+7. Open the aurora folder in Visual Studio Code
+
+# Testing #
+
+## Testing with molecule_docker ##
+
+Once you have written your code for aurora in your branch, test it locally with Molecule first before pushing to GitHub.
+
+1. In the docker container terminal go to /ansible/playbooks/molecule_docker folder
```
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: fixed

File: README.md [hunk 24]
```diff
@@ -1,8 +1,194 @@
...
+
+2. Run the following command to execute all molecule tests locally with debug mode:
+
+```
+molecule --debug test --all
+```
+
+3. Fix any errors
+
+4. If you want to run a specific test case (scenarios are in ansible/playbooks/molecule_docker/molecule folder), use:
```
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: changed wording to test case

File: README.md (hunk 17 above)
Comment (ToivoHn) [reply README.md]: done

File: README.md (hunk 9 above)
Comment (giusebar) [reply README.md]: agree

File: README.md (hunk 18 above)
Comment (ToivoHn) [reply README.md]: done

File: README.md (hunk 19 above)
Comment (ToivoHn) [reply README.md]: fixed

File: README.md (hunk 20 above)
Comment (ToivoHn) [reply README.md]: fixed

File: README.md [hunk 25]
```diff
@@ -1,8 +1,54 @@
...
+
+For example, it's possible to use Aurora to install Docker, download the specified image and create a new container for you. It will also create a desktop icon to start the container and launch the hand.
+
+Ansible user guide is available [here](https://docs.ansible.com/ansible/latest/user_guide/index.html) (Aurora is currently using Ansible 2.8.1)
+
+Molecule user guide is available [here](https://molecule.readthedocs.io/en/stable/) (Aurora is currently using Molecule 2.20.1)
+
+# How to run #
+
+## Ethercat interface ##
```
Comment (biofotis) [inline README.md]: this is not needed for teleop right? maybe it can be a subsection or hand_e and h products?

File: README.md (hunk 11 above)
Comment (ToivoHn) [reply README.md]: Done

File: README.md (hunk 25 above)
Comment (ToivoHn) [reply README.md]: fixed

File: README.md (hunk 10 above)
Comment (ToivoHn) [reply README.md]: Yes. I added a comment about it

File: README.md (hunk 9 above)
Comment (ToivoHn) [reply README.md]: resolved

File: README.md (hunk 16 above)
Comment (ToivoHn) [reply README.md]: resolved

File: README.md [hunk 26]
```diff
@@ -63,15 +77,659 @@ For assigning input and secure input to playbook variables you can use the tags:
...
+- name: Install Python 3
+  import_playbook: ./install_python3.yml
+
+- name: Tutorial 1 installation
+  hosts: docker_deploy
+  roles:
+    - {role: products/tutorial/deploy }
+```
+
+4. Create a role for in roles/products/tutorial folder (you will have to create the tutorial folder). Inside the tutorial folder, create the following folder structure and empty files where indicated. You should have the following file structure:
```
[the remark this answers is not in the document]
Comment (ToivoHn) [reply README.md]: done

**Pull request merged**